src/poll.rs

```diff
@@ -673,289 +1091,412 @@ impl ReadinessQueue {
         self.inner().awakener.wakeup()
     }
 
-    // Attempts to state to sleeping. This involves changing `head_readiness`
-    // to `sleep_token`. Returns true if `poll` can sleep.
-    fn prepare_for_sleep(&self) -> bool {
-        // Use relaxed as no memory besides the pointer is being sent across
-        // threads. Ordering doesn't matter, only the current value of
-        // `head_readiness`.
-        ptr::null_mut() == self.inner().head_readiness
-            .compare_and_swap(ptr::null_mut(), self.sleep_token(), Ordering::Relaxed)
-    }
-
-    fn take_ready(&self) -> ReadyList {
-        // Use `Acquire` ordering to ensure being able to read the latest
-        // values of all other atomic mutations.
-        let mut head = self.inner().head_readiness.swap(ptr::null_mut(), Ordering::Acquire);
-
-        if head == self.sleep_token() {
-            head = ptr::null_mut();
+    /// Prepend the given node to the head of the readiness queue. This is done
+    /// with relaxed ordering. Returns true if `Poll` needs to be woken up.
+    fn enqueue_node_with_wakeup(&self, node: &ReadinessNode) -> io::Result<()> {
+        if self.enqueue_node(node) {
+            try!(self.wakeup());
         }
 
-        ReadyList { head: ReadyRef::new(head) }
+        Ok(())
     }
 
-    fn new_readiness_node(&self, token: Token, interest: Ready, opts: PollOpt, ref_count: usize) -> ReadyRef {
-        let mut node = Box::new(ReadinessNode::new(token, interest, opts, ref_count));
-        let ret = ReadyRef::new(&mut *node as *mut ReadinessNode);
+    fn enqueue_node(&self, node: &ReadinessNode) -> bool {
+        let inner = self.inner();
+        let node_ptr = node as *const _ as *mut _;
 
-        node.next_all_nodes = self.inner_mut().head_all_nodes.take();
+        // Relaxed used as the ordering is "released" when swapping
+        // `head_readiness`
+        node.next_readiness.store(ptr::null_mut(), Relaxed);
 
-        let ptr = &*node as *const ReadinessNode as *mut ReadinessNode;
+        unsafe {
+            let prev = inner.head_readiness.swap(node_ptr, AcqRel);
 
-        if let Some(ref mut next) = node.next_all_nodes {
-            next.prev_all_nodes = ReadyRef::new(ptr);
-        }
+            debug_assert!((*prev).next_readiness.load(Relaxed).is_null());
 
-        self.inner_mut().head_all_nodes = Some(node);
+            (*prev).next_readiness.store(node_ptr, Release);
 
-        ret
+            prev == self.sleep_marker()
+        }
     }
 
-    /// Prepend the given node to the head of the readiness queue. This is done
-    /// with relaxed ordering. Returns true if `Poll` needs to be woken up.
-    fn prepend_readiness_node(&self, mut node: ReadyRef) -> bool {
-        let mut curr_head = self.inner().head_readiness.load(Ordering::Relaxed);
+    /// Must only be called in `poll`
```

Comment:
Could you mention here as well that this is the 1024 cores pop?

I'd also be fine if you wanted to make this fn unsafe and document that it's only safe to call from `poll`